src/utils/loose-equal.js

```diff
@@ -28,39 +28,35 @@ function looseEqual(a, b) {
   if (a === b) {
     return true
   }
-  if (typeof a !== typeof b) {
-    return false
+  const validDatesCount = [isDate(a), isDate(b)].filter(v => v === true).length
+  if (validDatesCount > 1) {
+    return validDatesCount === 2 ? a.getTime() === b.getTime() : false
   }
-  if (isDate(a) && isDate(b)) {
-    return a.getTime() === b.getTime()
+  const validFilesCount = [isFile(a), isFile(b)].filter(v => v === true).length
+  if (validFilesCount > 1) {
+    return validFilesCount === 2 ? a === b : false
   }
-  if (isFile(a) && isFile(b)) {
-    return a === b
+  const validArraysCount = [isArray(a), isArray(b)].filter(v => v === true).length
+  if (validArraysCount > 1) {
+    return validArraysCount === 2
+      ? a.length === b.length && a.every((e, i) => looseEqual(e, b[i]))
+      : false
   }
-  if (isArray(a) && isArray(b)) {
-    return a.length === b.length && a.every((e, i) => looseEqual(e, b[i]))
-  }
-  if (isObject(a) && isObject(b)) {
-    if (keys(a).length !== keys(b).length) {
+  const validObjectsCount = [isObject(a), isObject(b)].filter(v => v === true).length
+  if (validObjectsCount > 1) {
+    if (validObjectsCount === 1 || keys(a).length !== keys(b).length) {
```

Comment:
`validObjectsCount === 1` will never be true as the wrapping if is ` validObjectsCount > 1`